runtime/parachains/src/paras.rs

```diff
@@ -524,9 +665,174 @@ pub mod pallet {
 			Self::deposit_event(Event::ActionQueued(para, next_session));
 			Ok(())
 		}
+
+		/// Includes a statement for a PVF pre-checking vote. Potentially, finalizes the vote and
+		/// enacts the results if that was the last vote before achieving the supermajority.
+		#[pallet::weight(0)]
+		pub fn include_pvf_check_statement(
+			origin: OriginFor<T>,
+			stmt: PvfCheckStatement,
+			signature: ValidatorSignature,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+			let validators = shared::Pallet::<T>::active_validator_keys();
+			let current_session = shared::Pallet::<T>::session_index();
+			if stmt.session_index < current_session {
+				return Err(Error::<T>::PvfCheckStatementStale.into())
+			} else if stmt.session_index > current_session {
+				return Err(Error::<T>::PvfCheckStatementFuture.into())
+			}
+			let validator_index = stmt.validator_index.0 as usize;
+			let validator_public = validators
+				.get(validator_index)
+				.ok_or(Error::<T>::PvfCheckValidatorIndexOutOfBounds)?;
+
+			let signing_payload = stmt.signing_payload();
+			ensure!(
+				signature.verify(&signing_payload[..], &validator_public),
+				Error::<T>::PvfCheckInvalidSignature,
+			);
+
+			let mut active_vote = PvfActiveVoteMap::<T>::get(&stmt.subject)
```

Comment:
nit: Is there a particular reason we don't `try_mutate` this in place? It feels a bit error prone having the vote tracking the vote at the very end in an `else` case.

Reply:
yeah it was also bothering me, but I think in this case cure is the same or worse than the disease. 

In the case of achieving the quorum, we should remove the vote from the list and the map. Probably we can switch the entire storage value to be `OptionQuery`, but then we will have to deal with this everywhere as well. Then, I am trying to put mutations of the map closer to the list, but in this case the mutation will happen inside of `try_mutate` but the list mutation will be left alone.

We could perhaps refactor the code so that we return an enum, but that's additional code only to make sure we do not accidentally miss this update, so it feels to me not worth it